### PR TITLE
add Attribute class

### PR DESCRIPTION
**Pull request by superstar54**
Patch to #102 

The `_attribute`  is added to the Batoms object. It stores the properties of the attributes.

```python
>>> from ase.build import bulk
>>> import numpy as np
>>> from batoms import Batoms
>>> au = bulk('Au')
>>> d0 = np.zeros((len(au)))
>>> d2 = np.zeros((len(au), 2))
>>> d22 = np.zeros((len(au), 2, 2))
>>> au.set_array("d0d", d0)
>>> au.set_array("d1d", d2)
>>> au.set_array("d2d", d22)
>>> au._attributes
Name                Type      Domain    Dimension    Shape               
select              INT       POINT              0  [] 
species_index       INT       POINT              0  [] 
species             STRING    POINT              0  [] 
show                BOOLEAN   POINT              0  [] 
scale               FLOAT     POINT              0  [] 
model_style         INT       POINT              0  [] 
numbers             INT       POINT              0  [] 
d0d                 FLOAT     POINT              0  [] 
d1d                 FLOAT     POINT              1  [  2  ] 
d2d                 FLOAT     POINT              2  [  2    2  ] 
------------------------------------------------------------

>>> # Get attribute
>>> au.get_attribute('d2d').shape
(1, 2, 2)

>>> au.attributes['d2d'].shape
(1, 2, 2)

```

Performance
============

Time used to set and get attributes for a system with1000000 Au atoms

|Shape of array | Scatter (set) time | Gather (get) time | Comment
|---|---|---|---|
|  (1x1)   |     1.6    |    0.06    |  This is a normal data, as the reference |
| (1x2)    |     2.1    |   0.12     |  |
| (2x2)    |     3.7    |  0.24    | |

The performance looks good. The gather time scales linearly.

### CODE REVIEW

**Comment by alchem0x2A** [comment]
two edge case tests are added to `test_batoms.py`

**Comment by superstar54** [comment]
Thanks for pointing out the "attrib name conflict". This is the extreme case because the attribute read from a structure file usually has a meaningful name, e.g. velocity, stress, tensor and so on.

To avoid the conflict, two approaches come to my mind. 

1) Add a UUID to the name of each attribute, making the name unique globally. However, this would make the spreadsheet very ugly.

2) Check the name of the attribute before adding a new one. If name conflict exists, add the "_" symbol (recursively) to make the name unique.

I prefer the second approach, which keeps the spreadsheet readable.

**Comment by alchem0x2A** [comment]
agreed. solution 2 is better as I also don't expect normal user will mess around these attributes. 

Since the edge cases are only involved when the attribute name ends with a digit, inserting a delimiter between the name and index seems to be working. I'm thinking to use `@` as the delimiter in this case since it won't be occupied by ase-generated objects, after that the geometry nodes looks like this 
image

File: batoms/attribute.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            self[data["name"]] = data
+            att = self[data["name"]]
+            mesh = self.parent.obj.data
+            if data['dimension'] == 0:
+                mesh.attributes.new(name=data["name"],
+                        type=att.type, domain=att.domain)
+            else:
+                M = np.product(data["shape"][0:data["dimension"]])
+                for i in range(M):
+                    name = "{}{}".format(data["name"], i)
```
[the remark this answers is not in the document]
Comment (alchem0x2A) [reply batoms/attribute.py]: should be resolved from [7befeb0](https://github.com/superstar54/beautiful-atoms/pull/107/commits/7befeb0b094a927f5f6dd53513dbc1f37039d3d4) onwards.

File: batoms/attribute.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        Returns:
+            _type_: _description_
+        """
+        from batoms.utils import type_py_to_blender
+        # print("from_array: ", name, data)
+        array = np.asarray(data)
+        type_py = type(array.flat[0])
+        # print(name, dtype)
+        # Update the subdtype check according to 
+        # https://github.com/numpy/numpy/blob/db481babcfa7ebc70833e77985858e9295a3135b/numpy/core/numerictypes.py#L357
```
[the remark this answers is not in the document]
Comment (alchem0x2A) [reply batoms/attribute.py]: should be resolved from [bae8f02](https://github.com/superstar54/beautiful-atoms/pull/107/commits/bae8f02513f59db2d61ffd965638f4176cd8ef1c) onwards

**Comment by superstar54** [comment]
Yes, we can use @ as the delimiter.

To avoid the name conflict like "d@0" and "d". I define the delimiter as a variable and update the delimiter (delimiter += "@" recursively) when needed.

**Comment by alchem0x2A** [comment]
LGTM!
